java/src/jni/h5aImp.c

```diff
@@ -958,10 +1029,443 @@ Java_hdf_hdf5lib_H5_H5Awrite_1string(JNIEnv *env, jclass clss, jlong attr_id, jl
 /*
  * Class:     hdf_hdf5lib_H5
  * Method:    H5AreadVL
- * Signature: (JJ[Ljava/lang/String;)I
+ * Signature: (JJ[java/util/ArrayList;)I
  */
 JNIEXPORT jint JNICALL
 Java_hdf_hdf5lib_H5_H5AreadVL(JNIEnv *env, jclass clss, jlong attr_id, jlong mem_type_id, jobjectArray buf)
+{
+    H5T_class_t type_class;
+    hsize_t     dims[H5S_MAX_RANK];
+    hid_t       sid = H5I_INVALID_HID;
+    jsize       n;
+    htri_t      vl_data_class;
+    herr_t      status = FAIL;
+    jboolean    readBufIsCopy;
+    jbyteArray *readBuf = NULL;
+
+    UNUSED(clss);
+
+    if (NULL == buf)
+        H5_NULL_ARGUMENT_ERROR(ENVONLY, "H5AreadVL: read buffer is NULL");
+
+    /* Get size of data array */
+    if ((n = ENVPTR->GetArrayLength(ENVONLY, buf)) < 0) {
+        CHECK_JNI_EXCEPTION(ENVONLY, JNI_TRUE);
+        H5_BAD_ARGUMENT_ERROR(ENVONLY, "H5AreadVL: readBuf length < 0");
+    }
+
+    dims[0] = (hsize_t)n;
+    if ((sid = H5Screate_simple(1, dims, NULL)) < 0)
+        H5_LIBRARY_ERROR(ENVONLY);
+
+    if ((vl_data_class = h5str_detect_vlen(mem_type_id)) < 0)
+        H5_LIBRARY_ERROR(ENVONLY);
+
+    if ((type_class = H5Tget_class((hid_t)mem_type_id)) < 0)
+        H5_LIBRARY_ERROR(ENVONLY);
+    if (type_class == H5T_VLEN) {
+        size_t      typeSize;
+        hid_t       memb = H5I_INVALID_HID;
+        H5T_class_t vlClass;
+        size_t      vlSize;
+        void *      rawBuf = NULL;
+        jobject *   jList  = NULL;
+
+        size_t i, j, x;
+        char * cp_vp = NULL;
+
+        if (!(typeSize = H5Tget_size(mem_type_id)))
+            H5_LIBRARY_ERROR(ENVONLY);
+
+        if (!(memb = H5Tget_super(mem_type_id)))
+            H5_LIBRARY_ERROR(ENVONLY);
+        if ((vlClass = H5Tget_class((hid_t)memb)) < 0)
+            H5_LIBRARY_ERROR(ENVONLY);
+        if (!(vlSize = H5Tget_size(memb)))
+            H5_LIBRARY_ERROR(ENVONLY);
+
+        if (NULL == (rawBuf = HDcalloc((size_t)n, typeSize)))
```

Comment:
Does this ever get freed?

Reply:
Probably not - didn't check for that after getting it working.